### DISCUSSION TITLE
Eglot integration (snippet expansion)

### DISCUSSION THREAD

**Discussion by KaranAhlawat** [company-mode/company-mode#1414] [Q&A] [answered]
While using `corfu`, and having `yas-minor-mode` enabled, I was able to jump precisely through the fields of the snippet provided by `eglot` (the language server, to be more precise). However, now that I'm trying out `company-mode`, when I autocomplete some LSP suggestion that is also a snippet, the cursor just ends up skipping over the snippet's field. 

![image](https://github.com/company-mode/company-mode/assets/56113511/96831c12-87cd-4318-8b25-d53921d08f0f)

as seen in the image.
Is there some way to fix that?

**Comment by KaranAhlawat** [comment]
1. I do not have TNG mode enabled.
2. The language server is `metals`.
3. I was unable to find a fix for any similar issues. (or even similar issues, [un]fortunately).

**Reply by dgutov** [reply]
Indeed, expansion of completions using Eglot and Yasnippet has been known to work fine until now.

**Reply by dgutov** [reply]
@KaranAhlawat Could you try a similar scenario together with some other language server? E.g. `clangd`, it's usually easy to install.

**Reply by KaranAhlawat** [reply]
Sure, but I'm not entirely sure how to trigger snippets in `clangd`... Any clues on what snippet I should test?

I tested the `return expr` snippet, while it highlights the inserted `expression` text, typing doesn't replace it, only prepends at the start of it.

**Reply by dgutov** [reply]
`reinterpret_cast`, for instance.

**Reply by KaranAhlawat** [reply]
It half worked. In the first placeholder, it did what you'd expect it to. In the second one however, it doesn't really replace the text (although the placeholders aren't just $0 etc in this case).

**Reply by dgutov** [reply]
I think the second placeholder isn't supposed to be shown in that case -- it's just the last position for point.

Here's a video with me expanding that snippet, and with using a snippet expansion in a function call too: https://imgur.com/a/nItyrqG

**Reply by KaranAhlawat** [reply]
For me it goes like this -

[Screencast from 2023-10-16 23-39-26.webm](https://github.com/company-mode/company-mode/assets/56113511/eee37a78-9a73-4a32-b561-a4dc06dd1384)

And the other one goes like tihs (what prompted me to make the issue) -
[Screencast from 2023-10-16 23-40-58.webm](https://github.com/company-mode/company-mode/assets/56113511/809f2376-a561-4f54-950c-77fd0275b25b)

**Reply by dgutov** [reply]
Yeah, the text `expression` shouldn't stay around.

You can try upgrading to the latest version of eglot/company/yasnippet, and if it doesn't help.

Or try reproducing the problem starting with `emacs -Q` (to verify the absence of any conflicting settings and packages).

Failing all that, Eglot has specific instructions for reporting a bug (listing all current versions et cetera).

**Reply by joaotavora** [reply]
I couldn't reproduce: https://github.com/joaotavora/eglot/discussions/1308#discussioncomment-7299222

**Comment by KaranAhlawat** [comment]
I tried with `emacs -Q`, still running into the same issue. And I somehow broke my PGTK build upgrading my packages XD

**Reply by dgutov** [reply]
Can you try starting with the empty init directory as well? E.g. call `HOME=/tmp emacs`, then install the packages (tell us which versions), and see if that works better.

**Reply by dgutov** [reply]
Speaking of PGTK, though. It's very unlikely to be the cause of the problem, but JFYI it's only recommended if you are running on Wayland.

**Reply by KaranAhlawat** [reply]
I did run using `emacs -Q --init-directory=/tmp/emacs-test`.
The package versions are:
1. Eglot - 1.12.29
2. Company - 20231016.3
3. yas - 20200604.246

**Reply by KaranAhlawat** [reply]
As an aside, yes, I run on Wayland. PGTK was working fine till now, but updating all the packages broke something, but that's not related to this, so I think we can leave it at that

**Reply by dgutov** [reply]
And your Emacs is 29.1, I suppose? I think I've tried this particular combination of packages too, but I'll try again.

Can you try upgrading Eglot to 1.15? It's in the packages menu, but you'd have to "install" it.

**Reply by dgutov** [reply]
Something I noticed: right after you jumped to the second field, in the first video, was when the popup with the error description appeared in the right corner (for the first time). And then you typed a character, and the field became inactive/disappeared. What is that package?

Something like `flycheck-posframe`? I'm assuming you tested without it installed as well?

**Reply by KaranAhlawat** [reply]
I updated to `eglot` 1.15, did not change anything for me unfortunately.
That package is `eldoc-box`, and yes, I did test without when I did `emacs -Q`.

**Reply by dgutov** [reply]
I suppose a really far-out hypothesis could be that the PGTK build is interfering somehow. Do you build Emacs from source?

Any chance you can switch to `lucid` (the result should run okay through XWayland) and compare?

**Reply by dgutov** [reply]
And you're sure that when using Corfu you didn't have any such problems?

**Reply by KaranAhlawat** [reply]
> I suppose a really far-out hypothesis could be that the PGTK build is interfering somehow. Do you build Emacs from source?
> 
> Any chance you can switch to `lucid` (the result should run okay through XWayland) and compare?

Yes, I do build from source (debian repos are on 28). I don't see how it'd interfere, but I can find some time to build a `lucid` build.

> And you're sure that when using Corfu you didn't have any such problems?

Well, I went back to Corfu after our discussions yesterday, and it was doing the same thing now, so, it did have those problems as well I suppose 😄 Although I now think it's a Yasnippet thing rather than corfu or company.

I can 100% say `corfu` and `yasnippet` worked as expected *before* I switched to `company`, because I remember using that feature just the day before while working on a Scala project.

**Reply by KaranAhlawat** [reply]
Doesn't look like I'm the only one who ran into this either [as per this issue](https://github.com/joaotavora/yasnippet/issues/1141). As at least we know it's not entirely unique, although the `symbol-overlay` package isn't used by me so that solution was a dead-end.

**Reply by dgutov** [reply]
That issue mentions Emacs 28.1, and there was no pgtk back then :/

**Comment by dgutov** [comment]
Tangentially from the bug, it seems like `$0` is not so useful in these snippets.

For example, `return` expands into `return {$0:expression};`, meaning at the end the point remains before the semicolon. That seems less useful than being able to TAB again moving point to eol and pressing RET to open the next line.

But VS Code behaves this too, so I suppose this is a choice made by the language servers (at least clangd).

**Reply by KaranAhlawat** [reply]
Yeah I think it's actually a decision that iirc was decided in the Microsoft languag-server-protocol repository.

**Reply by dgutov** [reply]
Was there a discussion when to use the `$0` placeholder? As opposed to `$1`, etc.

**Reply by KaranAhlawat** [reply]
Ah no, looking it up, it was about something else (signature help while using snippets), sorry.

**Comment by KaranAhlawat** [comment]
Okay! Lots of new information, which will hopefully put this issue to rest.

1. Upgrading to clangd-15 also gives me the same completions as @joaotavora, with the last tab stop as `$0` instead of `${0:expression}`.
2. Apparently for `metals`, the issue was me modify the client capabilities and setting `dynamicRegistration` to true? Or maybe in general just modifying the client capabilities (which shouldn't break, but maybe it's just the `dynamicRegistration` part). Anyhow, if I set that, the server doesn't include the `includeTextFormat` field in the completion item it sends back, and then `eglot` doesn't treat it as a snippet. Not setting it, however, does make the server indicate the completion is a snippet which `eglot` can handle properly.

So, I think the only "bug" would be that `${0:text}` doesn't replace the `text` part, as in https://github.com/joaotavora/yasnippet/issues/1141. But that is a `yasnippet`.

**Reply by joaotavora** [reply, marked as answer]
I can't reproduce that either.  I had asked you already if you can successfully expand the following in your `*scratch*` buffer after `M-x yas-minor-mode`:

```
(yas-expand-snippet "hello ${0:world}")
```

I can, flawlessly, in both Emacs 30 and Emacs 29.  The decision whether to completely delete the "world" after the first typed character is controlled by `delete-selection-mode` (defaults to nil, but I think most people set it to t).

**Reply by KaranAhlawat** [reply]
Alright, that works. Thank you both very much!

**Reply by dgutov** [reply]
Aha! Now I get it too.

@joaotavora Should we do something more about this issue? It came up before (another example: https://github.com/emacs-lsp/lsp-dart/issues/130). The starter packs like Doom help have their own workarounds.

I guess in the past having a default value for the `$0` stop wasn't very popular (there are no such examples in the `yasnippet-snippets` collection), but with snippets coming from LSP it now happens more. And there is no visual indication that this kind of "final" field is different from the other template fields (that it is, in fact, just active region after point). The wrong default value of `delete-selection-mode` affects what looks like the expected behavior.

We could add a mention to the docs. But most don't read them until they really have to. Maybe we should really change yasnippet's handling of `$0`? When there is a default text, it would add an overlay just like in other cases, with the only difference that the last TAB won't move point later. Curiously, TAB currently works as expected there (aside from saying "Indenting the region" in the echo area :stuck_out_tongue_closed_eyes:).

**Reply by joaotavora** [reply]
In my testing, even with `delete-selection-mode` set to `nil` I still see the region overlay, it's just not automatically deleted.  Which is the expected behavior.

I'm not sure I like the overlay idea.  It used to be an overlay but I more or less distinctly deciding it should just be a region.  My reasoning was that yasnippet has done its job by that point, the snippet has exited, no more navigation possible etc.  If there was this overlay we would need to clean it up somehow. The visual region machinery should take care of that.

**Reply by dgutov** [reply]
> In my testing, even with delete-selection-mode set to nil I still see the region overlay, it's just not automatically deleted. Which is the expected behavior.

The region overlay at the end looks just like any other template field, so the difference in behavior (when it happens) is puzzling. How long did we take to understand the source of this complaint here?

> My reasoning was that yasnippet has done its job by that point, the snippet has exited, no more navigation possible etc. If there was this overlay we would need to clean it up somehow.

We would clean it up like the others. Like the last overlay in a template without `$0`, e.g. `"hello ${1:beautiful} ${2:world}"`.

Again, here's a comment in that other issue from the DartCode author and Analysis Server developer: https://github.com/emacs-lsp/lsp-dart/issues/130#issuecomment-1082225423

Perhaps `$0` itself is special, in that the "point moved away" hooks are less likely to fire soon, so the deletion of the overlay is predicated on the user typing something (increasing the odds of something going wrong before that), but wherever that decision was made (some discussion, perhaps, or another bug report) seems worth revisiting.

Maybe ideally that variable's default would change, but that seems unlikely to happen in any near future.

**Reply by dgutov** [reply]
Finally, if we don't want to change yasnippet anymore, the LSP clients could translate the templates instead, e.g. change

```snippet
"reinterpret_cast<${1:type}>(${0:expression})"
```

into

```snippet
"reinterpret_cast<${1:type}>(${2:expression}$0)"
```

before passing it to yasnippet.

**Reply by joaotavora** [reply]
> The region overlay at the end looks just like any other template field, so the difference in behavior (when it happens) is puzzling. How long did we take to understand the source of this complaint here?

Well I still don't understand it, to be honest.  And if we're talking about "looks", then I'm OK with changing yasnippet's field overlay face.

And I'm also OK with changing yasnippet to automatically delete the text of the such complex exit markers if `delete-selection-mode` is nil.  Not even worth hiding behind a user option, but that could be done.

Changing Eglot to massage snippets is a no-no.  Especially in that way.  An exit marker is an exit marker, it's not a field.

**Reply by dgutov** [reply]
> Well I still don't understand it, to be honest. And if we're talking about "looks", then I'm OK with changing yasnippet's field overlay face.

When two things look the same but behave differently, it's a problem. Changing the face of the fields might help in diagnosing user bug reports in such cases (that's an improvement), but it would be even better to remove the cause for those reports being created.

> And I'm also OK with changing yasnippet to automatically delete the text of the such complex exit markers if delete-selection-mode is nil.

Delete when the user starts typing, yes?

> Changing Eglot to massage snippets is a no-no. Especially in that way. An exit marker is an exit marker, it's not a field.

But it's both, when it has a default content. The resulting template behaves the same, except it's not affected by delete-selection-mode=nil. And then only eglot/lsp-mode would be at any risk of a regression, not any other users of yasnippet.

Anyway, you probably know better which of these approaches can be implemented more easily and reliably.

**Reply by joaotavora** [reply]
> Delete when the user starts typing, yes?

No, if `delete-selection-mode` is nil, either don't insert that "text" in `${0:text}` at all when expanding the snippet, or auto-delete it when moving to the exit marker.

> Changing the face of the fields might help in diagnosing user bug reports in such cases (that's an improvement), but it would be even better to remove the cause for those reports being created.

This has been Yasnippet's behaviour for perhaps like 10 years or sth like that?  Better just document it.  If, as you say, people don't read the documentation, then we should strive to change that instead, perhaps by writing better documentation.  It's like Project, or Xref or Company.  Eglot is a gateway to these packages.  It's understandable that people report some bugs to what they perceive as the frontend, but that can't be a substitute for learning how these parts of Emacs work.

**Reply by dgutov** [reply]
> No, if delete-selection-mode is nil, either don't insert that "text" in ${0:text} at all when expanding the snippet, or auto-delete it when moving to the exit marker.

It's worth a try, given that the latest clangd removed the default text for the exit marker anyway. I hope we don't lose something useful this way, though.

>  It's like Project, or Xref or Company. Eglot is a gateway to these packages. It's understandable that people report some bugs to what they perceive as the frontend, but that can't be a substitute for learning how these parts of Emacs work.

I do change 10-year-old+ behaviors in Company sometimes.

Learning the how whatever X works is good, but not when everybody tells you they can't reproduce your issue because they all customized a certain var 20+ years ago away from its default.

Anyway, doing the "don't insert ... at all ... or auto-delete it when moving to" thing should be an improvement over the current situation.

**Reply by joaotavora** [reply]
> Anyway, doing the "don't insert ... at all ... or auto-delete it when moving to" thing should be an improvement over the current situation.

Yes, but only if delete-selection-mode is nil.

**Reply by dgutov** [reply]
Yep.
